django/utils/deprecation.py

```diff
@@ -83,6 +85,157 @@ def __new__(cls, name, bases, attrs):
+    The decorated function will map any positional arguments after the `*`
+    to the corresponding keyword arguments, but issue a deprecation warning.
+
+    Works on both functions and methods. To apply to a class constructor,
+    decorate its __init__() method.
+
+    Example: to deprecate passing option1 or option2 as posargs, change::
+        def some_func(request, option1, option2=True):
+            ...
+
+    to::
+        @deprecate_posargs(RemovedInDjangoNMWarning, moved=["option1", "option2"])
+        def some_func(request, *, option1, option2=True):
+            ...
+
+    `moved` must be a list of the formerly positional-or-keyword parameters
+    that are now keyword-only, in their original positional order.
+
+    After the deprecation period, remove the decorator (but keep the `*`).
+
+    Caution: during the deprecation period, you must not add or remove any
+    positional parameters. (But it's OK to re-order or add keyword-only params.)
+    """
+
+    def decorator(func):
+        if isinstance(func, type):
+            raise TypeError(
+                "@deprecate_posargs cannot be applied to a class."
+                " (Apply it to the __init__ method.)"
+            )
+
+        params = inspect.signature(func).parameters
+        num_by_kind = Counter(param.kind for param in params.values())
+
+        if num_by_kind[inspect.Parameter.VAR_POSITIONAL] > 0:
+            raise TypeError(
+                "@deprecate_posargs() cannot be used with variable positional `*args`."
+            )
+
+        num_positional_params = (
+            num_by_kind[inspect.Parameter.POSITIONAL_ONLY]
+            + num_by_kind[inspect.Parameter.POSITIONAL_OR_KEYWORD]
+        )
+        num_keyword_only_params = num_by_kind[inspect.Parameter.KEYWORD_ONLY]
+        if num_keyword_only_params < 1:
+            raise TypeError(
+                "@deprecate_posargs() requires at least one keyword-only parameter"
+                " (after a `*` entry in the parameters list)."
+            )
+        if any(
+            name not in params or params[name].kind != inspect.Parameter.KEYWORD_ONLY
+            for name in moved
+        ):
+            raise TypeError(
+                "@deprecate_posargs() `moved` names must"
+                " all be keyword-only parameters."
+            )
+
+        num_moveable_args = len(moved)
+        max_positional_args = num_positional_params + num_moveable_args
+
+        param_names = list(params.keys())
+
+        # The deprecation message logic needs to know if func is bound to a class.
+        # But because function decorators always receive plain, unbound functions,
+        # class information is not available here. (inspect.ismethod(func) will
+        # always return False, and func.__self__ will never be set.) This only
+        # affects message construction, not argument mapping logic, so it's fine
+        # to just look for the customary bound parameter names.
```

Comment:
I think there *is* a way to accurately detect decoration of bound functions, class names, etc., by reworking the decorator as a descriptor. (That would also correctly handle non-standard bound param names like `this` or `klass`, as well as unbound functions that take `self` or `cls` params.)

But I'm not at all convinced that would be worth the added complexity. Plus I'm pretty sure there's additional overhead with a descriptor.

Reply:
Indeed, [implementing as a descriptor](https://github.com/django/django/commit/7a3e1f8251f569a0c6bb36c26be4f935927e576a) better handles [some obscure cases](https://github.com/django/django/blob/7a3e1f8251f569a0c6bb36c26be4f935927e576a/tests/deprecation/test_deprecate_posargs.py#L393-L455), but at the cost of 75% more overhead on method calls, and requiring counterintuitive decorator ordering with static methods. Doesn't seem worth it.